[Clang][analyzer] replace Stmt* with ConstCFGElement in SymbolConjured (reland)

fangyi-zhou opened the pull request:
Closes #57270.

This PR changes the `Stmt *` field in `SymbolConjured` with
`CFGBlock::ConstCFGElementRef`. The motivation is that, when conjuring a
symbol, there might not always be a statement available, causing
information to be lost for conjured symbols, whereas the CFGElementRef
can always be provided at the callsite.

Following the idea, this PR changes callsites of functions to create
conjured symbols, and replaces them with appropriate `CFGElementRef`s.

There is a caveat at loop widening, where the correct location is the
CFG terminator (which is not an element and does not have a ref). In
this case, the first element in the block is passed as a location.

Previous PR #128251, Reverted at #137304.

fangyi-zhou committed:
[Clang][analyzer] Replace Stmt* with ConstCFGElementRef in SymbolConjured (#128251)

This PR changes the `Stmt *` field in `SymbolConjured` with
`CFGBlock::ConstCFGElementRef`. The motivation is that, when conjuring a
symbol, there might not always be a statement available, causing
information to be lost for conjured symbols, whereas the CFGElementRef
can always be provided at the callsite.

Following the idea, this PR changes callsites of functions to create
conjured symbols, and replaces them with appropriate `CFGElementRef`s.

Closes #57270

fangyi-zhou commented:
Range diff against previous PR: https://gist.github.com/fangyi-zhou/7d3a73a9b95f93755af3e823228c7a0d

fangyi-zhou commented:
@steakhal

fangyi-zhou commented:
gentle ping

steakhal commented:
Sorry about my availability. I barely have any focus time these days.
Could you please split your commit into: 1) the original commit we had to later revert, 2) the changes to fix it that would make it on par with what you have here now

This would allow me to review the affected parts, saving me precious time. I know you did a diff of the diffs, but that's not really helpful.

fangyi-zhou commented:
Hi @steakhal 

Sorry to hear that you're busy. The main change wrt the previous commits are
(1) Rebase past the refactor in #137182
(2) Fix the bug that caused the ASan failure (when we visit a CFG block entrance)

I try to split the commit into two, but I'll mark the change in (2) in the meantime

fangyi-zhou commented on clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:
```diff
@@ -2556,10 +2556,19 @@ void ExprEngine::processCFGBlockEntrance(const BlockEdge &L,
     ProgramStateRef WidenedState =
-        getWidenedLoopState(Pred->getState(), LCtx, BlockCount, Term);
+        getWidenedLoopState(Pred->getState(), LCtx, BlockCount,
+                            *nodeBuilder.getContext().getBlock()->ref_begin());
```
Main change here: instead of passing the current CFG element in engine state, we use the first element in the current block.

fangyi-zhou committed:
Fix ASan error

When processing an CFG entrance, the expression engine does not have a
CFG element reference set up, thus an Asan error would be triggered if
we use the stale value. Instead, we pass the first element in the block
which would be a workaround.

fangyi-zhou commented:
Gentle ping

steakhal reviewed (approved):
LGTM. Thanks for the ping.

gregbedwell commented:
We've observed a crash in clang since this commit when using `--analyze` in conjunction with `-ftime-trace`.  I've put the details in #139779.  Please could you take a look?
